designs/2018-relative-package-loading/README.md

```diff
@@ -0,0 +1,278 @@
+
+## Summary
+
+This feature would update ESLint to plugins and shareable configs relative to the location of the config files where they're referenced, rather than relative to the location of the running ESLint instance. As a result, configs would be able to specify plugins as their own dependencies, and ESLint users would see fewer confusing errors about missing plugins.
+
+## Motivation
+
+Currently, ESLint plugins and shareable configs are loaded from the location of the ESLint package, rather than the location of the config file where the plugins and configs are referenced. This leads to several problems:
+
+* The current behavior requires end users to manually install any plugins required by a shareable config. As a result, a shareable config can't add a plugin without requiring new manual installation steps from the end user. This greatly decreases the ergonomics of using custom rules from plugins in shareable configs, and results in increased pressure to add new rules and options to ESLint core.
+* The current behavior assumes that if a user installs a config/plugin and ESLint in the same project, then ESLint will be able to load that config/plugin. This relies on an implementation detail of how npm works rather than a specified behavior, which leads to problems when using other package management strategies, e.g. with `lerna`. (More details about this problem can be found in [eslint/eslint#10125](https://github.com/eslint/eslint/issues/10125).) It also creates problems when using ESLint as a non-top-level dependency with package management strategies that strictly enforce dependency validity with Yarn Plug 'n Play.
+* The current behavior leads to a large amount of confusion from users where ESLint behaves differently depending on whether it's installed "globally" with npm.
+
+A separate but related facet of ESLint's design currently requires plugin names to be globally unique, leading to some additional problems:
+
+* If two shareable configs depend on two different versions of the same plugin, the configs are entirely incompatible and can't be used together.
+* It's difficult to create features that load plugins from other locations (e.g. as proposed in [eslint/eslint#6237](https://github.com/eslint/eslint/issues/6237)), because the name of a plugin loaded from another location might conflict with the name of a plugin in `node_modules`.
+
+### Design goals of solution
+
+* A config author should be able to add or upgrade any plugin in their config, without requiring additional installation steps from end users that extend that config.
+* The end user should maintain the ability to override any configuration setting inherited from an extended config.
+* A config author should be able to extend any two other configs at the same time, and have ESLint lint their code successfully. (The two configs might advocate mutually-incompatible code styles, but that issue is out of scope for this proposal.)
```

Comment:
We currently have a [hardcoded limit](https://github.com/eslint/eslint/blob/62fd2b93448966331db3eb2dfbe4e1273eb032b2/lib/linter.js#L33) for how many autofixes happen before we stop trying to apply them. I think the same limit would work fine here. (In other words, it would work the same way as mutually-incompatible fixes that don't originate from the same rule.)